### PR TITLE
Clean up Tensor MakeWithValue impls

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
Comment (nilsvu) [inline tests/Unit/DataStructures/Tensor/Test_Tensor.cpp]: optional: name `structure` -> `used_for_size`

File: src/DataStructures/Tensor/Tensor.hpp
Comment (nilsvu) [inline src/DataStructures/Tensor/Tensor.hpp:593]: optional: name `T` -> `DataType`

File: src/DataStructures/Tensor/Tensor.hpp
Comment (nilsvu) [inline src/DataStructures/Tensor/Tensor.hpp]: <s>DataVectors</s> data vectors

(more occurrences below, where the type is now more general than `DataVector`)

File: src/DataStructures/Tensor/Tensor.hpp
Comment (nilsvu) [inline src/DataStructures/Tensor/Tensor.hpp]: Could you point to where this generic version is implemented? I would have expected the overload that creates a `double` from a `Tensor<double>` would be located in `Tensor.hpp` (so having this comment is useful).

File: src/Utilities/MakeWithValue.hpp
Comment (nilsvu) [inline src/Utilities/MakeWithValue.hpp]: I think in other places we just write `typename = std::nullptr_t`, omitting the name. I have no opinion on this though, just pointing it out.

**Review by nilsvu — DISMISSED** [review]
Please feel free to squash changes related to these comments right away. Since I'll be on vacation from Aug 13th, please don't wait for my re-approval.

File: src/DataStructures/Tensor/Tensor.hpp
Comment (nilsdeppe) [reply src/DataStructures/Tensor/Tensor.hpp]: I don't think if `data vectors` is actually better.  I'd be fine with `vectors` or to keep as-is.

File: src/DataStructures/Tensor/Tensor.hpp
Comment (nilsdeppe) [reply src/DataStructures/Tensor/Tensor.hpp:593]: `T` is standard in this use case in C++. I'm not against changing it, though.

File: src/Utilities/MakeWithValue.hpp
Comment (nilsdeppe) [reply src/Utilities/MakeWithValue.hpp]: I agree, unless there is a compiler bug or some other reason it be named :)

**Review by nilsdeppe — DISMISSED** [review]
Looks good. I have no problems with merging as-is. @wthrowe please let us know if you would like to make any of the suggested changes :)

**Comment by wthrowe** [comment]
I made all the suggested changes except T -> DataType.  I also fixed the clang-tidy warnings in Test_Variables.cpp.

I have not made clang-tidy changes to Test_Tensor.cpp, because they seem questionable to me and there should probably be at least a brief discussion of how to handle them.  (Some of them suggest removing instances from calls to static methods, which may go against our "static when possible" rule (which is itself, I think, from clang-tidy).  Others suggest `auto` -> `auto*` on types that are implementation defined and just happen to be pointers in libstdc++.)

**Review by nilsdeppe — COMMENTED** [review]
LGTM! I'm happy for you to squash

I agree that the clang-tidy issues should be revisited separately :) Thanks for making this fixes and updates, @wthrowe!

**Review by nilsvu — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by kidder** [comment]
as discussed above, ignoring clang-tidy errors in `Test_Tensor.cpp`

**Pull request merged**
